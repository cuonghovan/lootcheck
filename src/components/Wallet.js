import React from 'react';
import { connect } from 'react-redux';
import { setBalanceAction, depositAction, withdrawAction } from '../actions/balance';

export class Wallet extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      balance: undefined
    }
  }

  onBalanceChange = event => this.setState({balance: parseInt(event.target.value, 10)});  

  setBalance = () => this.props.setBalanceAction(this.state.balance);

  depositBalance = () => this.props.depositAction(this.state.balance);

  withdrawBalance = () => this.props.withdrawAction(this.state.balance);

  render() {
    return (
      <div>
        <h1 className='balance'>Balance: {this.props.balance}</h1>
        <br/>
        <input className='balance-input' onChange={this.onBalanceChange} />
        <button className='set-btn' onClick={this.setBalance}>Set</button>
        <button className='deposit-btn' onClick={this.depositBalance}>Deposit</button>
        <button className='withdraw-btn' onClick={this.withdrawBalance}>Withdraw</button>
      </div>
    )
  }
}

const mapDispatchToProps = {
  setBalanceAction,
  depositAction,
  withdrawAction
}

export default connect((state) => ({ balance: state.balance }), mapDispatchToProps)(Wallet);
