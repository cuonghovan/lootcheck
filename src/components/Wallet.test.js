import React from 'react';
import { shallow } from 'enzyme';
import { Wallet } from './Wallet';

describe('Wallet', () => {
  const setBalanceFn = jest.fn();
  const depositBalanceFn = jest.fn();
  const withdrawBalanceFn = jest.fn();
  const props = {
    balance: 10,
    setBalanceAction: setBalanceFn,
    depositAction: depositBalanceFn,
    withdrawAction: withdrawBalanceFn  
  };
  const wallet = shallow(<Wallet {...props} />);

  it('renders correctly', () => {
    expect(wallet).toMatchSnapshot();
  });

  it('shows the balance', () => {
    expect(wallet.find('.balance').text()).toEqual('Balance: 10');
  });

  it('container an input', () => {
    expect(wallet.find('.balance-input').exists()).toBe(true);
  });

  describe('when user enters balance', () => {
    let balance = '10';

    beforeEach(() => {
      wallet.find('.balance-input').simulate('change', { target: { value: parseInt(balance, 10)}});
    })

    it('changes the balance in local state', () => {
      expect(wallet.state().balance).toEqual(10);
    });

    describe('and user wants to set balance', () => {
      beforeEach(() => {
        wallet.find('.set-btn').simulate('click');
      })
  
      it('calls `setBalance` once', () => {
        expect(setBalanceFn).toBeCalledWith(parseInt(balance, 10));
      })
    });

    describe('and user wants to deposit balance', () => {
      beforeEach(() => {
        wallet.find('.deposit-btn').simulate('click');
      })

      it('calls `depositAction` once', () => {
        expect(depositBalanceFn).toBeCalledWith(parseInt(balance, 10));
      })
    });

    describe('and user wants to withdraw balance', () => {
      beforeEach(() => {
        wallet.find('.withdraw-btn').simulate('click');
      })

      it('calls `withdrawAction` once', () => {
        expect(withdrawBalanceFn).toBeCalledWith(parseInt(balance, 10));
      });
    })
  });
}); 
