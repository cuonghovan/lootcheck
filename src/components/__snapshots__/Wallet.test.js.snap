// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wallet renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Wallet
    balance={10}
    depositAction={[Function]}
    setBalanceAction={[Function]}
    withdrawAction={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1
          className="balance"
        >
          Balance: 
          10
        </h1>,
        <br />,
        <input
          className="balance-input"
          onChange={[Function]}
        />,
        <button
          className="set-btn"
          onClick={[Function]}
        >
          Set
        </button>,
        <button
          className="deposit-btn"
          onClick={[Function]}
        >
          Deposit
        </button>,
        <button
          className="withdraw-btn"
          onClick={[Function]}
        >
          Withdraw
        </button>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Balance: ",
            10,
          ],
          "className": "balance",
        },
        "ref": null,
        "rendered": Array [
          "Balance: ",
          10,
        ],
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "br",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "balance-input",
          "onChange": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Set",
          "className": "set-btn",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Set",
        "type": "button",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Deposit",
          "className": "deposit-btn",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Deposit",
        "type": "button",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Withdraw",
          "className": "withdraw-btn",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Withdraw",
        "type": "button",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1
            className="balance"
          >
            Balance: 
            10
          </h1>,
          <br />,
          <input
            className="balance-input"
            onChange={[Function]}
          />,
          <button
            className="set-btn"
            onClick={[Function]}
          >
            Set
          </button>,
          <button
            className="deposit-btn"
            onClick={[Function]}
          >
            Deposit
          </button>,
          <button
            className="withdraw-btn"
            onClick={[Function]}
          >
            Withdraw
          </button>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Balance: ",
              10,
            ],
            "className": "balance",
          },
          "ref": null,
          "rendered": Array [
            "Balance: ",
            10,
          ],
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "br",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "balance-input",
            "onChange": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Set",
            "className": "set-btn",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Set",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Deposit",
            "className": "deposit-btn",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Deposit",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Withdraw",
            "className": "withdraw-btn",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Withdraw",
          "type": "button",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
